### PR TITLE
GDB-8282 Add error handling when downloading sparql result is JSON/JSON-LD

### PR DESCRIPTION
**Pull request by yordanalexandrov**
## What
Add error handling when downloading sparql results in JSON/JSON-LD

## Why
Since the JSON/JSON-LD format can not be streamed, it is loaded first in GDB memory and then sent. This can cause an out of memory exception. OME is handled in gdb and an error is sent, which has to be handled properly

## How
- added alternative fetch with $http for JSON/JSON-LD downloads with error handling

### CODE REVIEW

File: src/js/angular/core/directives/queryeditor/query-editor.directive.js [hunk 1]
```diff
@@ -574,21 +586,66 @@ function queryEditorDirective($timeout, $location, toastr, $repositories, Sparql
...
                     toastr.warning($translate.instant('query.editor.query.results.mismatch'));
                 }
-                $('#wb-download-infer').val(scope.currentQuery.inference);
-                $('#wb-download-sameAs').val(scope.currentQuery.sameAs);
-                const auth = localStorage.getItem('com.ontotext.graphdb.auth');
-                if (auth) {
-                    $('#wb-auth-token').val(auth);
+                if (fetchWithErrorHandling) {
+                    // JSON and JSON-LD have to be fetched in memory before sending, so error handling must be introduced
+                    $http.get(apiUrl, {
```
Comment (svilenvelikov) [inline src/js/angular/core/directives/queryeditor/query-editor.directive.js:603]: Can you move this in some of the rest controllers?

File: src/js/angular/core/directives/queryeditor/query-editor.directive.js [hunk 2]
```diff
@@ -574,21 +586,66 @@ function queryEditorDirective($timeout, $location, toastr, $repositories, Sparql
             window.yasr = yasr;
             yasr.afterCopy = afterCopy;
             yasr.getQueryResultsAsFormat = function (downloadFormat) {
-                // Simple cross-browser download with a form
-                const $wbDownload = $('#wb-download');
-                $wbDownload.attr('action', 'repositories/' + $repositories.getActiveRepository());
-                $('#wb-download-query').val(scope.currentQuery.query);
-                if (window.editor.getValue() !== scope.currentQuery.query) {
+                const exportFormatWithErrorHandling = scope.exportFormats.filter((format) => format.name.includes('JSON'));
```
Comment (boyan-tonchev) [inline src/js/angular/core/directives/queryeditor/query-editor.directive.js:589]: IMO will better if you use "type" or "extension" instead "name". Unlikely but the "name" can be changed .

File: src/js/angular/core/directives/queryeditor/query-editor.directive.js (hunk 1 above)
Comment (yordanalexandrov) [reply src/js/angular/core/directives/queryeditor/query-editor.directive.js]: Moved it to RepositoriesRestService

File: src/js/angular/core/directives/queryeditor/query-editor.directive.js (hunk 2 above)
Comment (yordanalexandrov) [reply src/js/angular/core/directives/queryeditor/query-editor.directive.js]: Refactored that

**Comment by yordanalexandrov** [comment]
Jenkins retry build

File: src/js/angular/core/directives/queryeditor/query-editor.directive.js [hunk 3]
```diff
@@ -574,21 +583,47 @@ function queryEditorDirective($timeout, $location, toastr, $repositories, Sparql
...
+                    query: window.editor.getValue(),
+                    infer: scope.currentQuery.inference,
+                    sameAs: scope.currentQuery.sameAs,
+                    auth: localStorage.getItem('com.ontotext.graphdb.auth')
+                };
+                const accept = downloadFormat;
+
+                const fetchWithErrorHandling = exportTypesToFetchWithErrorHandling.includes(downloadFormat);
+                if (window.editor.getValue() !== queryParams.query) {
                     toastr.warning($translate.instant('query.editor.query.results.mismatch'));
```
[the remark this answers is not in the document]
Comment (yordanalexandrov) [reply src/js/angular/core/directives/queryeditor/query-editor.directive.js:601]: Yes, you are correct.
I will set `queryParams.query` with `scope.currentQuery.query` as it was before

**Review by boyan-tonchev — DISMISSED** [review]

**Comment by yordanalexandrov** [comment]
Jenkins retry build

File: src/js/angular/rest/repositories.rest.service.js [hunk 4]
```diff
@@ -141,4 +142,28 @@ function RepositoriesRestService($http) {
             }
         });
     }
+
+    function downloadResultsAsFile(apiUrl, queryParams, acceptHeader) {
```
Comment (svilenvelikov) [inline src/js/angular/rest/repositories.rest.service.js:146]: This is strange. Why not providing only the repository id and avoid hardcoding the endpoint in the controller?

Comment (yordanalexandrov) [reply src/js/angular/rest/repositories.rest.service.js:146]: Yes, no reason for this I just extracted the function. I will change the argument

**Review by svilenvelikov — APPROVED** [review]

File: src/js/angular/rest/repositories.rest.service.js [hunk 5]
```diff
@@ -84,7 +85,7 @@ function RepositoriesRestService($http) {
     }
 
     function getPrefix(repositoryId, params) {
-        return $http.post(`${REPOSITORIES_ENDPOINT}/${repositoryId}/prefix`, null, { params })
+        return $http.post(`${REPOSITORIES_ENDPOINT}/${repositoryId}/prefix`, null, {params});
```
Comment (boyan-tonchev) [inline src/js/angular/rest/repositories.rest.service.js:88]: REPOSITORIES_ENDPOINT has value 'res/repositories', the endpoint you need is 'repositories'. Now there is an exception.

Comment (svilenvelikov) [reply src/js/angular/rest/repositories.rest.service.js:88]: hmm why is this difference? Is that they actually override the yasqe's request?

Comment (yordanalexandrov) [reply src/js/angular/rest/repositories.rest.service.js:88]: You are correct and this is the wrong service for this function. It should be in the `RDF4JRepositoriesRestService` where are calls to the correct endpoint.

**Review by boyan-tonchev — APPROVED** [review]

File: src/js/angular/rest/repositories.rest.service.js (hunk 5 above)
Comment (yordanalexandrov) [reply src/js/angular/rest/repositories.rest.service.js:87]: @svilenvelikov the `rest/repositories` is the GDB repositiory controller which deals with repo management. It was my mistake to use this service. The one that deals with the repository data is `repositories/*` endpoint, which is used by `RDF4JRepositoriesRestService`

**Review by svilenvelikov — APPROVED** [review]

**Comment by yordanalexandrov** [comment]
Jenkins retry build

**Comment by yordanalexandrov** [comment]
Jenkins retry build

**Comment by yordanalexandrov** [comment]
Jenkins retry build

**Comment by yordanalexandrov** [comment]
Jenkins retry build

File: test-cypress/integration/explore/class.hierarchy.spec.js [hunk 6]
```diff
@@ -172,7 +172,7 @@ describe('Class hierarchy screen validation', () => {
         getReturnButton().should('be.visible').click();
     });
 
-    it('Test class-hierarchy for given graph', () => {
+    it.skip('Test class-hierarchy for given graph', () => {
```
Comment (boyan-tonchev) [inline test-cypress/integration/explore/class.hierarchy.spec.js:175]: Will be better if we move this test in flake suite, so we can run it locally.

Comment (yordanalexandrov) [reply test-cypress/integration/explore/class.hierarchy.spec.js:175]: Just a test I will revert it

File: src/js/angular/rest/rdf4j.repositories.rest.service.js [hunk 7]
```diff
@@ -115,4 +116,28 @@ function RDF4JRepositoriesRestService($http, $repositories, $translate) {
...
     }
+
+    function downloadResultsAsFile(repositoryId, queryParams, acceptHeader) {
+        return $http.get(`${REPOSITORIES_ENDPOINT}/${repositoryId}`, {
+            headers: {
+                accept: acceptHeader
+            },
+            params: queryParams,
+            responseType: "blob"
+        }).then(function (res) {
```
Comment (teodossidossev) [inline src/js/angular/rest/rdf4j.repositories.rest.service.js:127]: please add a catch clause with a proper toastr message

**Review by boyan-tonchev — APPROVED** [review]

File: src/js/angular/core/directives/queryeditor/query-editor.directive.js [hunk 8]
```diff
@@ -574,21 +583,46 @@ function queryEditorDirective($timeout, $location, toastr, $repositories, Sparql
...
+                            saveAs(data, filename);
+                        }).catch(function (res) {
+                        // data is received as blob
+                        res.data.text()
+                            .then((message) => {
+                                toastr.error(message, $translate.instant('common.error'));
+                            });
+                    });
+                } else {
+                    // Simple cross-browser download with a form
```
Comment (teodossidossev) [inline src/js/angular/core/directives/queryeditor/query-editor.directive.js:614]: The setting of `wb-download-` val is a good candidate for an util method. You can create one if is used somewhere else.

File: src/js/angular/rest/rdf4j.repositories.rest.service.js (hunk 7 above)
Comment (yordanalexandrov) [reply src/js/angular/rest/rdf4j.repositories.rest.service.js:127]: There is a catch where it is used. It does not enter here if there is an error.

**Review by teodossidossev — APPROVED** [review]

File: src/js/angular/core/directives/queryeditor/query-editor.directive.js (hunk 8 above)
Comment (yordanalexandrov) [reply src/js/angular/core/directives/queryeditor/query-editor.directive.js:614]: No sure if used anyware else,, won't do.

**Pull request merged**